packages/kinvey-datastore/src/repositories/network-repository.js

```diff
@@ -0,0 +1,99 @@
+import { KinveyRequest, RequestMethod, DeltaFetchRequest } from 'kinvey-request';
+
+import { Repository } from './repository';
+import { buildCollectionUrl, ensureArray } from '../utils';
+
+/**
+ * @typedef RequestOptions
+ * @property {Operation} operation
+ * @property {Object} options
+ * @property {Query} [options.query]
+ * @property {Query} [options.query]
+ * @property {Object} [options.aggregation]
+ * @property {Number} [options.timeout]
+ * @property {Object} [options.properties]
+ * @property {Boolean} [options.skipBL]
+ * @property {Boolean} [options.trace]
+ */
+
+export class NetworkRepository extends Repository {
+  read(collection, query, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.GET, null, query, null, options = {});
+    return this._makeHttpRequest(requestConfig, options.useDeltaFetch);
+  }
+
+  readById(collection, entityId, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.GET, null, null, entityId, options);
+    return this._makeHttpRequest(requestConfig);
+  }
+
+  create(collection, entities, options) {
+    return this._processBatch(collection, RequestMethod.POST, entities, null, options);
+  }
+
+  update(collection, entities, options) {
+    return this._processBatch(collection, RequestMethod.PUT, entities, options);
+  }
+
+  deleteById(collection, entityId, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.DELETE, null, null, entityId, options);
+    return this._makeHttpRequest(requestConfig);
+  }
+
+  delete(collection, query, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.DELETE, null, query, null, options);
+    return this._makeHttpRequest(requestConfig);
+  }
+
+  count(collection, query, options) {
+    const requestConfig = this._buildRequestConfig(collection, RequestMethod.GET, null, query, '_count', options);
+    return this._makeHttpRequest(requestConfig);
+  }
+
+  _processBatch(collection, method, entities, options) {
+    const isSingle = !Array.isArray(entities);
+    const requestPromises = ensureArray(entities).map((entity) => {
+      const id = method === RequestMethod.PUT ? entity._id : null; // TODO: this isn't great :)
```

Comment:
Do you mean the logic, the syntax, or both? :)
The logic looks suspect. We support customers setting their own `_id` value on new created entities. This code seems to erase it.

Reply:
I meant that it isn't apparent what is going on by reading it. "if it's a PUT, do it by id". And it has tripped you up, it seems.

It doesn't take away the ability to create your own ids. If you do an update, it will do a PUT to the URL, slash id (which is an upsert for the backend and it works). If you are doing a create, it does a POST with whatever you give it. It shouldn't be removing anything - what do you mean?

Reply:
I guess the code is confusing - from just reading it here, it looks like this piece of code will preserve an `id` *only* in the case of `PUT`, otherwise discard it. From your explanation, that seems to not be the case.

I'm ok leaving it as-is. Maybe add a comment describing what's going on.

Reply:
I think this confusion stems from looking at a PR/diff, rather than the actual code.

This code only builds the `KinveyRequest` config object. It doesn't alter the entities being sent to the backend, it just looks for the `_id` property on the entity when doing a put, in order to make sure the correct URL is formed.